### ISSUE TITLE
[no-array-method-this-argument] Ignore usages with method references

### ISSUE THREAD

**Issue by boris-petrov** [sindresorhus/eslint-plugin-unicorn#1384]
The `no-array-method-this-argument` is useful and makes sense, however in some cases I believe it should not warn. For example:

```js
value.map(x.someMethod, x);
```

Imagine that `someMethod` uses `this` so it should be correct when invoking it.

The rule complains about this usage but 1) removing the `thisArg` is wrong and 2) changing the code to `value.map(v => x.someMethod(v))` is uglier (and is not as "functional-style" as the other).

So I propose in this case the rule to not warn for such code - where a method reference is used as the first argument to `map`.

**Comment by sindresorhus** [comment]
You can do:

```js
value.map(x.someMethod.bind(x));
```

**Comment by boris-petrov** [comment]
Sure, but that's uglier and slower (as it creates a new function). Sorry, should have mentioned that.

**Comment by fisker** [comment]
It's intentionally to enforce use `.map(fn.bind(foo))` instead of `.map(fn, foo)`.

But maybe we can add an option to only check useless this, like bound function or arrow function?

**Comment by sindresorhus** [comment]
> Sure, but that's uglier and slower (as it creates a new function)

Regarding "ugly". That's highly subjective. I find it the opposite; using reusable language feature composition rather than a one-off parameter. You also haven't provided any proof that it's slower (micro-benchmarks don't count). I would be surprised if JS engines didn't optimize this case.

**Comment by boris-petrov** [comment]
@sindresorhus - you're right that it's subjective. That's why I prefer the other one. I'm also not sure about the performance - I could only guess that creating a bound function is slower than passing `this` explicitly. 

Anyway, that's my opinion and preference. An option to skip such cases where the first argument is a method reference (perhaps also a function name but I could argue that in that case a `bind` is better) would be nice. :)

**Comment by sindresorhus** [comment]
I'm not neutral on adding such option, but if added, it should be `false` by default.

**Comment by fregante** [comment]
Closing: Both solutions conflicts with `no-array-callback-reference` and that it's been open without complete agreement for years.

While I think it kind of makes sense, either disable both rules or use this to satisfy both:

```js
value.map(() => x.someMethod());
```
